# **Practice 08: Boolean Operators and More Subsetting**

#### Kelani Guyton ####

---

**IMPORTANT**:
- Before starting the lab, save a copy to your google drive by navigating to `File > Save a copy in Drive` <br>

<br>


**Instructions**: Complete the lab below.

- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- **NOTE: Your completed lab must be submitted to Canvas as a .ipynb file by the end of the lab session**
---

<br>

####<font color=red>Organized</font>

**Answers should be organized in the following format for organization and readability**


<font color=red>Example task 1</font>

In [1]:
# 1.

# This will be the code block to answer the first part of a task

1. This will be the text answer/explanation (if prompted) for the first task

<font color=red>Example task 2</font>

In [2]:
# 2.

# This will be the code block to answer the second part of a task

2. This will be the text answer/explanation (if prompted) for the second task

<br>

Continue this format for the remaining subtasks


<br>

---

<br>

#### Packages

In [3]:
# install packages for today's lecture
install.packages("dslabs")
install.packages("pheatmap")  # package for heatmap

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [4]:
# load the necessary packages
library(dslabs)
library(pheatmap)

<br>

<br>

<br>

---

####<font color=red>Task 1</font> - Loading and Preprocessing the Dataset



<font color=red>Task 1.1</font> Today, we'll be working with salary data available online.

* Run the following script to load the salary dataset for today's assignment.

```
# load data frame
data_url                       <- "https://raw.githubusercontent.com/khasenst/datasets_teaching/refs/heads/main/salary_data.csv"
salary                         <- na.omit(read.csv(data_url, header = TRUE))

# convert character variables to factors
salary$Gender                  <- as.factor(salary$Gender)
salary$Education.Level         <- as.factor(salary$Education.Level)
salary$Job.Title               <- as.factor(salary$Job.Title)

# reassign levels
levels(salary$Education.Level) <-  c("missing", "BS", "BS", "HS",
                                     "MS", "MS", "PhD", "PhD")
salary$Education.Level         <- factor(salary$Education.Level, levels = c("missing", "HS", "BS", "MS", "PhD"))
```


In [5]:
# load data frame
data_url                       <- "https://raw.githubusercontent.com/khasenst/datasets_teaching/refs/heads/main/salary_data.csv"
salary                         <- na.omit(read.csv(data_url, header = TRUE))

# convert character variables to factors
salary$Gender                  <- as.factor(salary$Gender)
salary$Education.Level         <- as.factor(salary$Education.Level)
salary$Job.Title               <- as.factor(salary$Job.Title)

# reassign levels
levels(salary$Education.Level) <-  c("missing", "BS", "BS", "HS",
                                     "MS", "MS", "PhD", "PhD")
salary$Education.Level         <- factor(salary$Education.Level, levels = c("missing", "HS", "BS", "MS", "PhD"))

<br>

<font color=red>Task 1.2</font> View the first few lines of the `salary` data frame using the `head()` function.

In [6]:
head(salary)

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
1,32,Male,BS,Software Engineer,5,90000
2,28,Female,MS,Data Analyst,3,65000
3,45,Male,PhD,Senior Manager,15,150000
4,36,Female,BS,Sales Associate,7,60000
5,52,Male,MS,Director,20,200000
6,29,Male,BS,Marketing Analyst,2,55000


<br>

<font color=red>Task 1.3</font> Use the `str()` function to determine the number of rows and columns in the `salary` data frame. Write your observations in a text cell.

In [7]:
str(salary)

'data.frame':	6699 obs. of  6 variables:
 $ Age                : int  32 28 45 36 52 29 42 31 26 38 ...
 $ Gender             : Factor w/ 3 levels "Female","Male",..: 2 1 2 1 2 2 1 2 1 2 ...
 $ Education.Level    : Factor w/ 5 levels "missing","HS",..: 3 4 5 3 4 3 4 3 3 5 ...
 $ Job.Title          : Factor w/ 192 levels "Account Manager",..: 177 19 146 117 27 95 108 120 96 167 ...
 $ Years.of.Experience: num  5 3 15 7 20 2 12 4 1 10 ...
 $ Salary             : int  90000 65000 150000 60000 200000 55000 120000 80000 45000 110000 ...
 - attr(*, "na.action")= 'omit' Named int [1:5] 173 261 3137 5248 6456
  ..- attr(*, "names")= chr [1:5] "173" "261" "3137" "5248" ...


There are 6699 observations(rows) and 6 columns (variables).

<br>

<br>

<br>

---

####<font color=red>Task 2</font> - Boolean Operators and Order of Execution



<font color=red>Task 2.1</font>

* What is the result of the following Boolean expression?

  `!FALSE | TRUE & FALSE`

* Explain the order in which the Booleans are evaluated in a text cell.

In [8]:
!FALSE | TRUE & FALSE

[1] TRUE

The `!FALSE` is first evaluated to `TRUE`, resulting in `TRUE | TRUE & FALSE`.

The AND operator is then evaluated, resulting in `TRUE | FALSE`.

The OR operator is then evaluated, resulting in `TRUE`.

<br>

<font color=red>Task 2.2</font>

* What is the result of the following Boolean expression?

  `(!FALSE | TRUE) & FALSE`

* Explain the order in which the Booleans are evaluated in a text cell.

In [9]:
(!FALSE | TRUE) & FALSE

[1] FALSE

The `!FALSE` is first evaluated to `TRUE`, resulting in `(TRUE|TRUE) & FALSE`.

The expression in the parenthesis is then evaluated, resulting in `TRUE & FALSE`.

The AND operator is then evaluated, resulting in `FALSE`.

<br>

<br>

<br>

---

####<font color=red>Task 3</font> - Subsetting with Boolean Operators



<font color=red>Task 3.1</font>

* What is the average salary of individuals who have a BS degree AND are less than 25 years of age?

* What is the average salary of individuals who have a PhD degree AND are less than 25 years of age?

* In your opinion, are these salaries very different? Why? Why not?

In [12]:
# have BS degree & less than 25 years old
mean(salary$Salary[salary$Education.Level == "BS" & salary$Age < 25])

# have a PhD degree & less than 25 years old
mean(salary$Salary[salary$Education.Level == "PhD" & salary$Age < 25])
length(salary$Salary[salary$Education.Level == "PhD" & salary$Age < 25])

[1] 60384.16

[1] 69875.31

[1] 16

The salaries differ by less than $10k..

<br>

<font color=red>Task 3.2</font>

* What is the average salary of individuals who have a BS degree AND are greater than or equal to 25 years of age?

* What is the average salary of individuals who have a PhD degree AND are greater than or equal to 25 years of age?

* In your opinion, are these salaries very different?

<br>

In [13]:
# have BS degree & greater than or equal to 25 years old
mean(salary$Salary[salary$Education.Level == "BS" & salary$Age >= 25])

# have a PhD degree & greater than or equal to 25 years old
mean(salary$Salary[salary$Education.Level == "PhD" & salary$Age >= 25])
length(salary$Salary[salary$Education.Level == "PhD" & salary$Age >= 25])

[1] 97622.14

[1] 166784.1

[1] 1353

These salaries are quite different due to the difference in education levls.

<font color=red>Task 3.3</font>

* How many individuals have either a MS OR a PhD degree?

* Use the NOT logical operator and your answer to the bullet above to determine how many individuals do NOT have an MS OR PhD degree?

In [14]:
# have MS or PhD
nrow(salary[salary$Education.Level == "MS" | salary$Education.Level == "PhD", ])

# no have MS or PhD
nrow(salary[!(salary$Education.Level == "MS" | salary$Education.Level == "PhD"), ])

[1] 3229

[1] 3470

<br>

<br>

<br>

---

####<font color=red>Task 4</font> - `which()`, `which.min()`, `which.max()`



<font color=red>Task 4.1</font>
* Use the `which()` function to determine the row indexes of individuals with the following criteria
  * Less than 30 years of age
  * Female
  * Salary greater than $175,000

* How many individuals in the data frame satisfy this criteria? Hint: You can use the `length()` function on the output of `which()`


In [15]:
which(salary$Age < 30 &                  # less that 30 years old
  salary$Gender == "Female" &            # gender is female
  salary$Salary > 175000)                # salary greater than $175,000

  # returns vector with row indexes that match these conditions

[1] 506 530 542 550 560 570 580 590 620 632 640 650 660 670 700 712 720 730 740
[20] 750 760 770 780 790 800 811 828 846 856 866 876 886 896 906 916 926 936 944
[39] 953

In [17]:
indexes_conditions_met <- which(salary$Age < 30 &                      # less that 30 years old
                               salary$Gender == "Female" &            # gender is female
                               salary$Salary > 175000)                # salary greater than $175,000

# pull out the rows of interest
salary[indexes_conditions_met, ]

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
508,29,Female,MS,Data Scientist,6,180000
532,29,Female,MS,Data Scientist,6,180000
544,29,Female,MS,Data Scientist,6,180000
552,29,Female,MS,Data Scientist,6,180000
562,29,Female,MS,Data Scientist,6,180000
572,29,Female,MS,Data Scientist,6,180000
582,29,Female,MS,Data Scientist,6,180000
592,29,Female,MS,Data Scientist,6,180000
622,29,Female,MS,Data Scientist,6,180000


In [18]:
length(which(salary$Age < 30 &                  # less that 30 years old
  salary$Gender == "Female" &                   # gender is female
  salary$Salary > 175000))                      # salary greater than $175,000

[1] 39

There are 39 individuals in the data frame that satisfy the specified criteria.

<br>

<font color=red>Task 4.2</font>
* Use the `which.min()` function to determine the index of the minimum salary

* Use the index above to extract the corresponding row of the data frame

* What is the job title for this individual?


In [20]:
which.min(salary$Salary)

salary[which.min(salary$Salary), ]

[1] 259

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
260,29,Male,BS,Junior Business Operations Analyst,1.5,350


The individual's job title is a Junior Business Operations Analyst.

<br>

<font color=red>Task 4.3</font>
* Use the `which.max()` function to determine the index of the maximum salary

* Use the index above to extract the corresponding row of the data frame

* What is the job title for this individual?


In [21]:
which.max(salary$Salary)

salary[which.max(salary$Salary), ]

[1] 31

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
31,50,Male,BS,CEO,25,250000


The individual's job title is a CEO.

<br>

<br>

<br>

---

####<font color=red>Task 5</font> - `order()`



<font color=red>Task 5.1</font>

* Order the `salary` data frame from smallest salary to largest salary using the `order()` function

* Store the ordered data frame as the variable `salary_low_high`

* Show the first 10 rows using the `head()` function

* In a text cell, write your observations of these 10 rows.

In [22]:
salary_low_high <- salary[order(salary$Salary), ]

# look at the first observations
head(salary_low_high, 10)

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
260,29,Male,BS,Junior Business Operations Analyst,1.5,350
4634,31,Female,BS,Junior HR Coordinator,4.0,500
1891,25,Female,BS,Front end Developer,1.0,550
2655,23,Male,PhD,Software Engineer Manager,1.0,579
3894,22,Male,HS,Junior Sales Associate,1.0,25000
3907,29,Female,HS,Junior Sales Associate,1.0,25000
3922,29,Female,HS,Junior Sales Associate,1.0,25000
3937,29,Female,HS,Junior Sales Associate,1.0,25000
3952,29,Female,HS,Junior Sales Associate,1.0,25000


The most common job title observed is "Junior Sales Associate" with few years of experience.

<br>

<font color=red>Task 5.2</font>

* Order the `salary` data frame from largest salary to smallest salary using the `order()` function

* Store the ordered data frame as the variable `salary_high_low`

* Show the first 10 rows using the `head()` function

* In a text cell, write your observations of these 10 rows.

In [23]:
salary_high_low <- salary[order(salary$Salary, decreasing = TRUE), ]

# look at the first observations
head(salary_high_low, 10)

,Age,Gender,Education.Level,Job.Title,Years.of.Experience,Salary
,<int>,<fct>,<fct>,<fct>,<dbl>,<int>
31,50,Male,BS,CEO,25,250000
84,52,Male,PhD,Chief Technology Officer,24,250000
5002,45,Male,BS,Financial Manager,21,250000
4277,51,Male,PhD,Data Scientist,24,240000
4292,51,Male,PhD,Data Scientist,24,240000
4307,51,Male,PhD,Data Scientist,24,240000
4322,51,Male,PhD,Data Scientist,24,240000
4337,51,Male,PhD,Data Scientist,24,240000
4352,51,Male,PhD,Data Scientist,24,240000


The most common job title is "Data Scientist" with greater than 20 years of experience. Most have a PhD and all are "Male".

<br>

<br>